src/backend/distributed/utils/citus_copyfuncs.c

```diff
@@ -0,0 +1,239 @@
+
+	COPY_SCALAR_FIELD(relationId);
+	COPY_SCALAR_FIELD(storageType);
+	COPY_SCALAR_FIELD(valueTypeId);
+	COPY_SCALAR_FIELD(valueTypeLen);
+	COPY_SCALAR_FIELD(valueByVal);
+	COPY_SCALAR_FIELD(minValueExists);
+	COPY_SCALAR_FIELD(maxValueExists);
+
+	if (!from->minValueExists)
+	{
+		newnode->minValue = datumCopy(from->minValue,
+									  from->valueByVal,
+									  from->valueTypeLen);
+	}
+	if (!from->maxValueExists)
+	{
+		newnode->maxValue = datumCopy(from->maxValue,
+									  from->valueByVal,
+									  from->valueTypeLen);
+	}
+
+	COPY_SCALAR_FIELD(shardId);
+}
+
+
+void
+CopyNodeMapMergeJob(COPYFUNC_ARGS)
+{
+	int arrayLength = 0;
+	int i = 0;
+
+	CITUS_COPY_LOCALS(MapMergeJob);
+
+	copyJobInfo(&newnode->job, &from->job);
+
+	COPY_NODE_FIELD(reduceQuery);
+	COPY_SCALAR_FIELD(partitionType);
+	COPY_NODE_FIELD(partitionColumn);
+	COPY_SCALAR_FIELD(partitionCount);
+	COPY_SCALAR_FIELD(sortedShardIntervalArrayLength);
+
+	arrayLength = from->sortedShardIntervalArrayLength;
+
+	/* now build & read sortedShardIntervalArray */
+	newnode->sortedShardIntervalArray =
+		(ShardInterval **) palloc(arrayLength * sizeof(ShardInterval *));
+
+	for (i = 0; i < arrayLength; ++i)
+	{
+		/* can't use COPY_NODE_FIELD, no field names */
+		newnode->sortedShardIntervalArray[i] =
+			copyObject(from->sortedShardIntervalArray[i]);
+	}
+
+	COPY_NODE_FIELD(mapTaskList);
+	COPY_NODE_FIELD(mergeTaskList);
+}
+
+
+void
+CopyNodeShardPlacement(COPYFUNC_ARGS)
+{
+	CITUS_COPY_LOCALS(ShardPlacement);
+
+	COPY_SCALAR_FIELD(placementId);
+	COPY_SCALAR_FIELD(shardId);
+	COPY_SCALAR_FIELD(shardLength);
+	COPY_SCALAR_FIELD(shardState);
+	COPY_STRING_FIELD(nodeName);
+	COPY_SCALAR_FIELD(nodePort);
+	COPY_SCALAR_FIELD(partitionMethod);
+	COPY_SCALAR_FIELD(colocationGroupId);
+	COPY_SCALAR_FIELD(representativeValue);
+}
+
+
+void
+CopyNodeRelationShard(COPYFUNC_ARGS)
+{
+	CITUS_COPY_LOCALS(RelationShard);
+
+	COPY_SCALAR_FIELD(relationId);
+	COPY_SCALAR_FIELD(shardId);
+}
+
+
+void
+CopyNodeTask(COPYFUNC_ARGS)
+{
+	CITUS_COPY_LOCALS(Task);
+
+	COPY_SCALAR_FIELD(taskType);
+	COPY_SCALAR_FIELD(jobId);
+	COPY_SCALAR_FIELD(taskId);
+	COPY_STRING_FIELD(queryString);
+	COPY_SCALAR_FIELD(anchorShardId);
+	COPY_NODE_FIELD(taskPlacementList);
+	COPY_NODE_FIELD(dependedTaskList);
+	COPY_SCALAR_FIELD(partitionId);
+	COPY_SCALAR_FIELD(upstreamTaskId);
+	COPY_NODE_FIELD(shardInterval);
+	COPY_SCALAR_FIELD(assignmentConstrained);
+	COPY_SCALAR_FIELD(shardId);
+	COPY_NODE_FIELD(taskExecution);
+	COPY_SCALAR_FIELD(upsertQuery);
+	COPY_SCALAR_FIELD(replicationModel);
+	COPY_SCALAR_FIELD(insertSelectQuery);
+	COPY_NODE_FIELD(relationShardList);
+}
+
+
+void
+CopyNodeDeferredErrorMessage(COPYFUNC_ARGS)
+{
+	CITUS_COPY_LOCALS(DeferredErrorMessage);
+
+	COPY_SCALAR_FIELD(code);
+	COPY_STRING_FIELD(message);
+	COPY_STRING_FIELD(detail);
+	COPY_STRING_FIELD(hint);
+	COPY_STRING_FIELD(filename);
+	COPY_SCALAR_FIELD(linenumber);
+	COPY_STRING_FIELD(functionname);
+}
```

Comment:
I wrote these in a few minutes, did you check whether I've got the fields to be copied right in these functions?